### PR TITLE
memdb engine supports foreign keys and cascade deletings&archiving&re…

### PR DESCRIPTION
**Pull request by savelyev-an**
support 
1) FK
2) cascade deletes 
3) cascade archiving
4) cascade restoring

### CODE REVIEW

File: vault-plugins/shared/memdb/memdb.go [hunk 1]
```diff
@@ -0,0 +1,407 @@
...
+	builder := strings.Builder{}
+	if len(msgs) != 0 {
+		for _, msg := range msgs {
+			if builder.Len() != 0 {
+				builder.WriteString(";")
+			}
+			builder.WriteString(msg)
+		}
+		return fmt.Errorf(builder.String()+":%w", relationHandlerError)
+	}
```
Comment (shvgn) [inline vault-plugins/shared/memdb/memdb.go:252]: I think we should have single common error collecting type for this. Formatting error messages is not a particular feature of a relation processing. (an example of the idea https://play.golang.org/p/eztQmlaleB from [SO](https://stackoverflow.com/questions/33470649/combine-multiple-error-strings))

File: vault-plugins/shared/memdb/memdb.go [hunk 2]
```diff
@@ -0,0 +1,407 @@
...
+
+func (t *Txn) checkCheckingRelations(table string, obj interface{}) error {
+	rels := t.schema.CheckingRelations[table]
+	return t.processRelations(rels, obj, t.checkRelationShouldBeEmpty, ErrNotEmptyRelation)
+}
+
+// implement main loop checking relations
+func (t *Txn) processRelations(relations []Relation, obj interface{},
+	relationHandler func(originObjectFieldValue interface{}, key Relation) error, relationHandlerError error) error {
+	valueIface := reflect.ValueOf(obj)
```
Comment (shvgn) [inline vault-plugins/shared/memdb/memdb.go:228]: The relationHandler type is so long, I didn't parse it right for the first time. Maybe we should give it a name

File: vault-plugins/shared/memdb/memdb.go [hunk 3]
```diff
@@ -0,0 +1,407 @@
...
+		err = t.CascadeDelete(key.RelatedDataType, relatedRecord)
+		if err != nil {
+			return fmt.Errorf("deleting related record: at table %q by index %q, value %q",
+				key.RelatedDataType, key.RelatedDataTypeFieldIndexName, parentObjectFiledValue)
+		}
+	}
+	return nil
+}
+
+func (t *Txn) archiveChildren(archivingTimeStamp int64, archivingHash int64) func(originObjectFieldValue interface{}, key Relation) error {
```
Comment (shvgn) [inline vault-plugins/shared/memdb/memdb.go:294]:
``` 
archivingTimeStamp int64, archivingHash int64
```

This dudes always come together. I'd make a dedicated struct for this pair. 

```go
type ArchiveMark struct {
  TimeStamp, Hash int64
}
```

**Review by shvgn — COMMENTED** [review]
Looks like a power feature.

File: vault-plugins/shared/memdb/dbschema.go [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+					key.RelatedDataTypeFieldIndexName, key, d)
+			}
+			ks[key] = struct{}{}
+		}
+		rels[d] = ks
+	}
+	for d := range rels {
+		err := validateCyclic(d, rels)
+		if err != nil {
+			return fmt.Errorf("cyclic rependency: %s",
```
[the remark this answers is not in the document]
Comment (savelyev-an) [reply vault-plugins/shared/memdb/dbschema.go:98]: fix

File: vault-plugins/shared/memdb/dbschema.go [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+}
+
+// validateCyclic checks absence of cyclic dependencies between tables
+func validateCyclic(topDataType dataType, rels map[dataType]map[Relation]struct{}) error {
+	type dependency struct {
+		parentType     dataType
+		curIdx         int
+		childDataTypes []dataType
+	}
+	childDataTypesFunc := func(parentDataType dataType, rels map[dataType]map[Relation]struct{}) []dataType {
```
[the remark this answers is not in the document]
Comment (savelyev-an) [reply vault-plugins/shared/memdb/dbschema.go:112]: fix

File: vault-plugins/shared/memdb/dbschema.go [hunk 6]
```diff
@@ -0,0 +1,291 @@
...
+	}
+	deps := make([]dependency, len(rels)+1)
+	deps[0] = dependency{
+		parentType:     topDataType,
+		curIdx:         0,
+		childDataTypes: childDataTypesFunc(topDataType, rels),
+	}
+	l := 0
+loop:
+	for deps[l].curIdx < len(deps[l].childDataTypes) || l != 0 {
```
[the remark this answers is not in the document]
Comment (savelyev-an) [reply vault-plugins/shared/memdb/dbschema.go:134]: fixed

File: vault-plugins/shared/memdb/dbschema.go [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+		return result
+	}
+	deps := make([]dependency, len(rels)+1)
+	deps[0] = dependency{
+		parentType:     topDataType,
+		curIdx:         0,
+		childDataTypes: childDataTypesFunc(topDataType, rels),
+	}
+	l := 0
+loop:
```
[the remark this answers is not in the document]
Comment (savelyev-an) [reply vault-plugins/shared/memdb/dbschema.go:133]: fixed

File: vault-plugins/shared/memdb/dbschema.go [hunk 8]
```diff
@@ -0,0 +1,291 @@
...
+			chainBuilder.WriteString("=>" + deps[0].parentType)
+			return fmt.Errorf("dependencies chain:%s", chainBuilder.String())
+		case len(rels[deps[l].childDataTypes[curIdx]]) > 0:
+			curType := deps[l].childDataTypes[curIdx]
+			l++
+			deps[l] = dependency{
+				parentType:     curType,
+				curIdx:         0,
+				childDataTypes: childDataTypesFunc(curType, rels),
+			}
```
[the remark this answers is not in the document]
Comment (savelyev-an) [reply vault-plugins/shared/memdb/dbschema.go:159]: fixed

File: vault-plugins/shared/memdb/memdb.go (hunk 3 above)
Comment (savelyev-an) [reply vault-plugins/shared/memdb/memdb.go:294]: TODO

File: vault-plugins/shared/memdb/dbschema.go [hunk 9]
```diff
@@ -0,0 +1,291 @@
...
+					} else {
+						return fmt.Errorf("index named %q not found at table %q, passed as relation to field %q of table %q",
+							r.RelatedDataTypeFieldIndexName, r.RelatedDataType, r.OriginalDataTypeFieldName, dt)
+					}
+				}
+			}
+			(*mapRels)[dt] = rs
+		}
+		return nil
+	}
```
[the remark this answers is not in the document]
Comment (savelyev-an) [reply vault-plugins/shared/memdb/dbschema.go:198]: fixed

File: vault-plugins/shared/memdb/memdb.go (hunk 2 above)
Comment (savelyev-an) [reply vault-plugins/shared/memdb/memdb.go:228]: I give a new line for this argument

File: vault-plugins/shared/memdb/dbschema.go [hunk 10]
```diff
@@ -0,0 +1,291 @@
...
+			// create chain
+			chainBuilder := strings.Builder{}
+			for i := 0; i <= l; i++ {
+				if chainBuilder.Len() != 0 {
+					chainBuilder.WriteString("=>")
+				}
+				chainBuilder.WriteString(deps[i].parentType)
+			}
+			chainBuilder.WriteString("=>" + deps[0].parentType)
+			return fmt.Errorf("dependencies chain:%s", chainBuilder.String())
```
[the remark this answers is not in the document]
Comment (savelyev-an) [reply vault-plugins/shared/memdb/dbschema.go:151]: fixed

File: vault-plugins/shared/memdb/memdb.go (hunk 1 above)
Comment (savelyev-an) [reply vault-plugins/shared/memdb/memdb.go:252]: fixed

**Pull request merged**